lib/manager/flux/extract.ts

```diff
@@ -0,0 +1,116 @@
+import { loadAll } from 'js-yaml';
+import { HelmDatasource } from '../../datasource/helm';
+import { logger } from '../../logger';
+import { SkipReason } from '../../types';
+import { readLocalFile } from '../../util/fs';
+import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
+import type {
+  FluxManifest,
+  FluxResource,
+  HelmRelease,
+  HelmRepository,
+} from './types';
+
+function readManifest(content: string): FluxManifest {
+  const manifest: FluxManifest = { releases: [], repositories: [] };
+  let resources: FluxResource[];
+  try {
+    resources = loadAll(content, null, { json: true }) as FluxResource[];
+  } catch (err) {
+    logger.debug({ err }, 'Failed to parse Flux manifest');
+    return null;
+  }
+
+  // It's possible there are other non-Flux HelmRelease/HelmRepository CRs out there, so we filter based on apiVersion.
+  for (const resource of resources) {
+    switch (resource.kind) {
+      case 'HelmRelease':
+        if (resource.apiVersion.startsWith('helm.toolkit.fluxcd.io/')) {
+          manifest.releases.push(resource);
+        }
+        break;
+      case 'HelmRepository':
+        if (resource.apiVersion.startsWith('source.toolkit.fluxcd.io/')) {
+          manifest.repositories.push(resource);
+        }
+        break;
+    }
+  }
+
+  return manifest;
+}
+
+function resolveReleases(
+  releases: HelmRelease[],
+  repositories: HelmRepository[]
+): PackageDependency[] {
+  return releases.map((release) => {
+    const res: PackageDependency = {
+      depName: release.spec.chart.spec.chart,
+      currentValue: release.spec.chart.spec.version,
```

Comment:
These need some defensive checks (`?.`) unless you have pre-validated them to exist

Reply:
Interesting. I had assumed that TS would validate objects during a type cast, but that was a misunderstanding - and come to think of it, that would be a lot of code for the compiler to generate. I'll fix this.

Reply:
Added these checks along with a slew of additional tests. This exercise reminded me of both the good and the bad of the years I spent practicing TDD... 🙂